In [12]:
import numpy as np
import pandas as pd

In [13]:
def approx_prio(judg_df):
    total_sum = judg_df.sum().sum()
    return judg_df.sum(axis=1) / total_sum

def eigen_prio(judg_df):
    eig_val, eig_vect = np.linalg.eig(judg_df.as_matrix())
    priorities = eig_vect[:,0]
    # There is no imaginary part, it is mainly a type conversion
    return np.real(priorities / priorities.sum())

def inconsistency(judg_df):
    eig_val, eig_vect = np.linalg.eig(judg_df.as_matrix())
    n = judg_df.shape[0]
    # consistency index
    index = (np.real(eig_val[0]) - n) / (n - 1)
    # priority vector
    w = eig_vect[:,0]
    w = np.real(w / w.sum())
    D = np.diag(w)
    E = np.dot(np.linalg.solve(D,judg_df), D)
    return index, E
    

# Decision making with the analytic hierarchy process
**Reference**: Int. J. Services Sciences, Vol. 1, No. 1, 2008

**Author**: Thomas L. Saaty

## Relative consumption of drinks

Here is the *judgment matrix* for the relative consumption of drinks in the US.

In [14]:
comp_mat = np.array([
[1,9,5,2,1,1,1/2], 
[1/9,1,1/3,1/9,1/9,1/9,1/9],
[1/5,2,1,1/3,1/4,1/3,1/9],
[1/2,9,3,1,1/2,1,1/3],
[1,9,4,2,1,2,1/2],
[1,9,3,1,1/2,1,1/3],
[2,9,9,3,2,3,1]])
labels = ['coffee','wine','tea','beer','sodas','milk','water']
df = pd.DataFrame(comp_mat, index = labels, columns=labels)
df

coffee  wine       tea      beer     sodas      milk     water
coffee  1.000000   9.0  5.000000  2.000000  1.000000  1.000000  0.500000
wine    0.111111   1.0  0.333333  0.111111  0.111111  0.111111  0.111111
tea     0.200000   2.0  1.000000  0.333333  0.250000  0.333333  0.111111
beer    0.500000   9.0  3.000000  1.000000  0.500000  1.000000  0.333333
sodas   1.000000   9.0  4.000000  2.000000  1.000000  2.000000  0.500000
milk    1.000000   9.0  3.000000  1.000000  0.500000  1.000000  0.333333
water   2.000000   9.0  9.000000  3.000000  2.000000  3.000000  1.000000

### Approximate calculus of priorities
The *priorities* are computed as the sum of each row divided by the total sum of the *comparison* or *judgment matrix*:

In [15]:
approx_prio(df)

coffee    0.185215
wine      0.017941
tea       0.040156
beer      0.145639
sodas     0.185215
milk      0.150388
water     0.275447
dtype: float64

### Accurate Calculus of priorities
The priorities are calculted from the eigen value/vector problem. The eigen vector is then normalized by the sum of the priority vector components. As it is positive real valued, it is equivalent to a normalization by the L1-norm.

In [16]:
eigen_prio(df).round(3)

array([ 0.178,  0.019,  0.039,  0.117,  0.19 ,  0.129,  0.327])

# Relative measurement and its generalization in decision making
# Why pairwise comparisons are central in mathematics for the measurement of intangible factors
# The Analytic Hierarchy/Network process
**Reference**: Rev. R. Acad. Cien. Serie A. Mat. VOL. 102(2), 2008, pp. 251-318

**Author**: Thomas L. Saaty

## Judgments, outcomes, and actual relative sizes of the five geometric shapes

In [17]:
judg_mat = np.array([
[1,9,2,3,5],
[1/9,1,1/5,1/3,1/2],
[1/2,5,1,3/2,3],
[1/3,3,2/3,1,3/2],
[1/5,2,1/3,2/3,1]])
labels = ['circle','triangle','square','diamond','rectangle']
df = pd.DataFrame(judg_mat, index = labels, columns=labels)
df

circle  triangle    square   diamond  rectangle
circle     1.000000       9.0  2.000000  3.000000        5.0
triangle   0.111111       1.0  0.200000  0.333333        0.5
square     0.500000       5.0  1.000000  1.500000        3.0
diamond    0.333333       3.0  0.666667  1.000000        1.5
rectangle  0.200000       2.0  0.333333  0.666667        1.0

### Calculating priorities
The *priorities* are computed as the sum of each row divided by the total sum of the *comparison* or *judgment matrix*:

In [18]:
approx_prio(df)

circle       0.456158
triangle     0.048910
square       0.250887
diamond      0.148251
rectangle    0.095793
dtype: float64

Accurate computation of priorities:

In [19]:
eigen_prio(df).round(3)

array([ 0.462,  0.049,  0.245,  0.151,  0.093])

In [23]:
ind, mat_E = inconsistency(df)
print('The consistency index is:',ind,'\n')
print('The inconsitency matrix is:')
print(mat_E,'\n')
print('The consistency deviation matrix:')
print(mat_E-1,'\n')
print("The maximal deviation is",(mat_E - 1).max().max())

The consistency index is: 0.00193921173303 

The inconsitency matrix is:
[[ 1.          0.95868012  1.0615159   0.97987755  1.00768328]
 [ 1.0431008   1.          0.99654128  1.02211106  0.94600371]
 [ 0.94204901  1.00347073  1.          0.92309268  1.13914444]
 [ 1.02053568  0.97836726  1.08331485  1.          0.92553906]
 [ 0.99237531  1.05707831  0.87785181  1.08045143  1.        ]] 

The consistency deviation matrix:
[[ 0.         -0.04131988  0.0615159  -0.02012245  0.00768328]
 [ 0.0431008   0.         -0.00345872  0.02211106 -0.05399629]
 [-0.05795099  0.00347073  0.         -0.07690732  0.13914444]
 [ 0.02053568 -0.02163274  0.08331485  0.         -0.07446094]
 [-0.00762469  0.05707831 -0.12214819  0.08045143  0.        ]] 

The maximal deviation is 0.139144436666


The statistically maximal admissible *consistency index* is 0.10 >> 0.0019. The matrix is very consistent.